### ISSUE TITLE
ProducesResponseTypeMetadata(200) added automatically to endpoint metadata in minimal apis

### ISSUE THREAD

**Issue by nicro950** [dotnet/aspnetcore#57801]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

It looks like minimal apis adds a ProducesResponseTypeMetadata with status 200 automatically to some Map*(Get/Put/Post/...) configurations. From my testing it looks like it happens as long as the Map* is given a delegate which returns some data, including Task or Task_T. In some cases this might be what you want, but if the endpoint, for instance, only returns a 201 Created, then the endpoint metadata is reporting that it both return a 200 and 201(As long as you also adds the Produces(201) metadata), which it does not do. This became a problem for us further down the line, when we tried to generate the corresponding OpenApi document for the metadata information, which added the extra response code, and set the response type to an empty object.

As seen in the image, both endpoint which returns a Task, have the extra metadata, but the last one returning only void does not. I also tried returning a string, and then it added the 200, but that would be the correct behavior in that context. The 200 response was removed when returning an Result.NoContent() directly without having it in a Task. 

![image](https://github.com/user-attachments/assets/8063a446-a8a1-4ba9-bac6-685a28ec6bf1)

### Expected Behavior

Either that ProducesResponseTypeMetadata is never added automatically or that it is only automatically added if the endpoint is guaranteed to have a scenario which results in a 200 status code. Looks like the intention was that it gets the 200 if the endpoint returns a type, but it should not get the 200 if it returns void, IResult, Task, or Task_IResult but the last two was not included.

### Steps To Reproduce

1. Create a new asp.net core application with minimal apis
2. Add the following endpoints:
```csharp
        app.MapPost("/test", () =>
            {
                return Task.CompletedTask;
            });

        app.MapPost("/test2", () =>
            {
                return Task.FromResult(Results.NoContent);
            })
            .Produces(201);

        app.MapPost("/test3", () =>
            {
            });
```
3. The first two will have the extra metadata added to them.

### .NET Version

9.0.100-rc.1.24452.12

**Comment by captainsafia** [comment]
@nicro950 Thanks for filing this issue!

Some thoughts on the scenarios that you have shared here.

`() => Task.CompletedTask` returning a 200 description _is_ expected behavior. By default, all void-returning methods will have `200 OK` description associated with them because that matches the framework behavior here. You'll observe that your last `test3` example works the same way.

`() => Task.FromResult(Results.NoContent)` is a bit of an interesting scenario. In this particularly case, you're not returning a `NoContent` object, you're returning the _function_ that returns a `NoContent` object. I think what you're probably trying to do here is something like `() => Task.FromResult(TypedResults.Created(...))`. 

Let me know if this response is helpful or if you're seeing other issues with this!

**Comment by nicro950** [comment]
@captainsafia  I can agree with the Task automatically adding 200, but then, should not the 3. option also add the 200 ProducesResponseTypeMetadata, which returns void? 

The code I wrote here was the simplest form of our problem, showing only the tings resulting in the extra ProducesResponseTypeMetadata being added to the metadata. Our problem is spesific for file upload which returns NoContent after the file is uploaded, and is more closely related to the following code:

```csharp
        app.MapPost("/file", async (IFormFile file) =>
        {
            await FileUpload(file);
            return Results.NoContent;
        })
        .Produces(StatusCodes.Status204NoContent);
```
where it gets both the 200 and 204, even if the endpoint can never return 200.

![image](https://github.com/user-attachments/assets/c63c75f1-c93a-4525-a948-bfea34a1ddeb)

**Comment by martincostello** [comment]
`return Results.NoContent;` is still returning the function itself, not its result.

Try:

```diff
- return Results.NoContent;
+ return Results.NoContent();
```

**Comment by nicro950** [comment]
@martincostello Sorry about that, but it does not look like it is related to the problem even if you return the object result instead of, it still adds both of the statuses, since our original code do work the proper way, but by trying to remove the most, resulted in in dissaparing:
![image](https://github.com/user-attachments/assets/3acfb950-2bdd-4233-99c4-7737dc2022d8)

**Comment by nicro950** [comment]
@martincostello @captainsafia here should be a proper example, showing all the types involved (so i can get them correct myself as well 😅). As we can see, both the 200 and 204 are added to the metadata list. Sorry about the bad examples. 

![image](https://github.com/user-attachments/assets/cd474e17-4e92-48c2-93f1-55a1de20212e)

**Comment by captainsafia** [comment]
> As we can see, both the 200 and 204 are added to the metadata list. Sorry about the bad examples.

I think the problem that you are seeing here is related to the fact that you are using `Results.NoContent` instead of `TypedResults.NoContent`. The `Results.NoContent` type doesn't produce any metadata about what it produces to the API. In lieu of this, minimal APIs will make the same assumption that it makes about `void`-returning types: that they should return 200 OK.

Updating your API to use `TypedResults.NoContent` should get you the right behavior here.

**Comment by nicro950** [comment]
Using TypedResult, and it still adds the 200 return status:
![image](https://github.com/user-attachments/assets/6b0e61cd-d5e5-4acd-bd02-a1d9099beede)

That logic would also break the moment you return multiple statuses if you have some not found check, or conflict check or something like that, and the function needs to fallback to using IResult as the common return type.

**Comment by chrisoverzero** [comment]
> That logic would also break the moment you return multiple statuses […]

This is what [the composite Results return type][composite] is meant for.

[composite]: https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/responses?view=aspnetcore-8.0#resultstresult1-tresultn

**Comment by captainsafia** [comment]
> @chrisoverzero Nice, did not now about that one, since we have mostly abstracted that layer away, but it still adds the 200 in that case 😅 ![image](https://private-user-images.githubusercontent.com/1964338/366892321-65ff27d5-da16-44e1-bd8b-193da1e67e05.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxNjYyMDgsIm5iZiI6MTcyNjE2NTkwOCwicGF0aCI6Ii8xOTY0MzM4LzM2Njg5MjMyMS02NWZmMjdkNS1kYTE2LTQ0ZTEtYmQ4Yi0xOTNkYTFlNjdlMDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDkxMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA5MTJUMTgzMTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmNlODAyZWU0OWFmNzZlYmU4NDQ3MmFmZTM2MWZiN2Y0MGZkMTg4ODlkOWIxZDA5Zjc3MGZjN2E3OWZkODcyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ViCIyS4Vi7EuJbwU-Une6PX35AfUM7biDjyGL3Uw-cQ)
> 
> Personally, I just feel like if a IResult is returned (Either IResult or Task_IResult), then it should not make any assumption about what HTTP status is returned from the result, and it should be up to the user to add it to the metadata manually.

Minor nit: the typical pattern is to use the `Results<>` type as the return for the method instead of casting the given case.

The `ProducesResponseTypeMetadata` that gets added here is a result of the recent work that we've done to support native AoT for minimal APIs + OpenAPI. Essentially, we use that metadata as a replacement for dynamic reflection for resolving the return type from methods that return awaitable types (like yours).

When this metadata is processed by the ApiExplorer layer, it gets filters out from the ApiResponse objects that are constructed. IMO, for realizing what response types are actually produced by an endpoint API explorer (and not endpoint metadata) should be the source of truth.

The original proposal for resolving this bug proposed using a new metadata type to capture await able unwrapping but we opted to use `ProducesResponseType` since ti already exists. We can consider adding a property to indicate that it is auto-generated.

Is there a particular library that is querying this metadata and running into issues?

**Comment by nicro950** [comment]
> Minor nit: the typical pattern is to use the Results<> type as the return for the method instead of casting the given case.

Yeah, I know, it was just the easiest way to force the return type resolution to chose the correct type, instead of having to create a separate function. 

> The ProducesResponseTypeMetadata that gets added here is a result of the recent work that we've done to support native AoT for minimal APIs + OpenAPI. Essentially, we use that metadata as a replacement for dynamic reflection for resolving the return type from methods that return awaitable types (like yours).

Aha, that make sense, I went in to take a look at the code, and the code worked as i expected before it was changed in a pr which had something to do with Native AoT, which I have little knowledge about, so I trust that you do the right thing there.

The reason I made the bug in that way in the first place was that the metadata was the first place i looked when open api generator started generating 200 statuses with result type void, in the form of `SwaggerResponse<void>` which did not make sense at all. Then afterwards I manually removed the 200 from the endpoint metadata and the problem was resolved in our code.

So a little more context, we are using NSwag to generate both the OpenAPI document, as well as C# clients from the OpenAPI document afterwards. I went digging into the NSwag source code with stepthrough debugging and it looks like NSwag uses the `Microsoft.OpenApi.Models.OpenApiOperation` added from `.WithOpenApi()` with the `Microsoft.AspNetCore.OpenApi` package with version `9.0.0-rc.1.24452.1`.

While inspecting the code it looks like it does not filter out the 200 status code:
![image](https://github.com/user-attachments/assets/67166f11-b90c-45f9-95b1-6c5542e90620)

Which like you mentioned that ApiExplorer does:
![image](https://github.com/user-attachments/assets/cff3185b-00e4-4784-9d7c-33bb50746ec8)

Not quite sure if this issue should be edited and renamed, or moved somewhere else, or created again in another repo, or if there is an intention to change status code resolution of the `.WithOpenApi()`?

> The original proposal for resolving this bug proposed using a new metadata type to capture await able unwrapping but we opted to use ProducesResponseType since ti already exists. We can consider adding a property to indicate that it is auto-generated. 

I was at least confused by the automatically added status code, which i never added, and probably some other code out there which might depend on the minimal api metadata. But I think that you have a larger knowledge context on the matter, so would be hard for me to know about all the nuances with choosing one option or another here.

**Comment by captainsafia** [comment]
> So a little more context, we are using NSwag to generate both the OpenAPI document, as well as C# clients from the OpenAPI document afterwards. I went digging into the NSwag source code with stepthrough debugging and it looks like NSwag uses the Microsoft.OpenApi.Models.OpenApiOperation added from .WithOpenApi() with the Microsoft.AspNetCore.OpenApi package with version 9.0.0-rc.1.24452.1.

Oh! It's rather surprising to me that NSwag is using the `OpenApiOperation` in metadata that is generated by the `WithOpenApi` extension method. I filed this issue in the NSwag repo (https://github.com/RicoSuter/NSwag/issues/4163) back when we first introduced the `WithOpenApi` extension method in .NET 7. As of today, the issue appears to be open so I assumed that NSwag didn't support the `WithOpenApi` strategy.

Can you verify what version of NSwag you're running into this issue with?

**Comment by nicro950** [comment]
We are using the latest version of nswag 14.1.0 in our project. I think this is the code that fetches the OpenApiOperation and uses that: https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L241

**Comment by captainsafia** [comment]
> Aha, interesting. We are using the latest version of nswag 14.1.0 in our project. I think this is the code that fetches the OpenApiOperation and uses that: https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L241, but not sure if it uses all of the functionality of the metadata or just the operations it produces.

Thanks for digging this up! I just took a look at the code and here's my interpretation of how it works. The fundamental issue here is the fact that unlike `Microsoft.AspNetCore.OpenApi` and `Swashbuckle` which both use the types in the [Microsoft.OpenApi package](https://www.nuget.org/packages/Microsoft.OpenApi) to describe an OpenAPI document, NSwag uses its own types to represent the OpenAPI document. So, NSwag pulls the `OpenApiOperation` from metadata ([1](https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L238-L241)), resolves the serializer for that type from the `Microsoft.OpenApi` library using private reflection ([2](https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L246)), uses the serializer to serialize that operation to a string ([3](https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L249-L250)), then deserializes that string using its _own_ serializer APIs to the its _own_ types ([4](https://github.com/RicoSuter/NSwag/blob/7d6df3af95081f3f0ed6dee04be8d27faa86f91a/src/NSwag.Generation.AspNetCore/AspNetCoreOpenApiDocumentGenerator.cs#L252)).

TL;DR: it uses private reflection + string-based round tripping to map between the types. There's a chance that there is a bug in this round-tripping logic that means some properties don't get fully round-tripped?

I re-opened this issue because I thought there might be something here but I think the best thing to do in this case is to file a bug on NSwag about this issue.

**Comment by nicro950** [comment]
@captainsafia Found that Swashbuckle version `6.7.3` exhibits the same behavior when using the WithOpenApi Endpoint metadata:

![image](https://github.com/user-attachments/assets/37a0595b-2522-464a-8ded-67ca98142c97)
And then without it:
![image](https://github.com/user-attachments/assets/d3be0574-8d65-4766-9458-7dd4f6967578)

So it probably does that since the `WithOpenApi` metadata from the `Microsoft.AspNetCore.OpenApi` package, adds the `OpenApiOperation` in the endpoint metadata which says that there are two statuses instead of one: 
![image](https://github.com/user-attachments/assets/6a7e0253-c00e-426b-904b-c7631c20b2d0)

**Comment by captainsafia** [comment]
@nicro950 Good debugging! I realize that we need to update [the logic here as well](https://github.com/dotnet/aspnetcore/blob/542f2b9877da67e90cca6ce1b65aa9557fbb2aa8/src/OpenApi/src/Services/OpenApiGenerator.cs#L112-L133) 🤦🏽‍♀️ 

I'll reopen this issue and see if it is possible to get a fix out for .NET 9 GA.

**Comment by nicro950** [comment]
@captainsafia No problem, happy to help, and sounds good :)

**Comment by dj-nitehawk** [comment]
@captainsafia 
i believe there's a bug/regression introduced in .net9.0's `ApiResponseTypeProvider` which is possibly related to this issue.
here's the simplest repro:
```cs
var bld = WebApplication.CreateBuilder(args);
bld.Services
   .AddEndpointsApiExplorer()
   .AddOpenApiDocument(); // using nswag

var app = bld.Build();

app.MapGet("test", () => new MyResponse { Content = "hello!" })
   .Produces<MyResponse>(200, "application/json"); //this is ignored if response type implements IResult

app.UseOpenApi()
   .UseSwaggerUi();
app.Run();

public class MyResponse : IResult //commenting out IResult or switching to net8.0 sdk fixes it
{
    public required string Content { get; set; }

    public Task ExecuteAsync(HttpContext httpContext)
        => throw new NotImplementedException();
}
```
causes nswag to spit out an incorrect `responses` section like this:
![Image](https://github.com/user-attachments/assets/4d19c2b3-ccf3-4e7b-814c-761ffbf8851a)

switching the project sdk version to `net8.0` or removing `IResult` from the response class brings back the correct swagger spec:
![Image](https://github.com/user-attachments/assets/94bc01d3-01ce-41e6-9f31-91d4413859cd)

the situation doesn't change even if the delegate is changed to this:

```cs
app.MapGet("test", () => TypedResults.Ok(new MyResponse { Content = "hello!" }))
```

is ignoring custom `ProducesResponseTypeMetadata` added to endpoints an intentional breaking change in .net 9?

thanks!

**Comment by captainsafia** [comment]
@dj-nitehawk This isn't an intentional breaking change as it is a known issue. I unfortunately didn't have the time to post a PR with a bug fix for this before our 9.0 GA so this will have to catch the 9.1 wave.

As you mentioned, this issue is specific to the use of `IResult` as the `Type` property in `ProducesResponseTypeMetadata`. In .NET 9, we made a chance to support native AoT that re-used `ProducesResponseTypeMetadata`. The bug is specifically a result of [this bit of code in the ApiResponseTypeProvider](https://source.dot.net/#Microsoft.AspNetCore.Mvc.ApiExplorer/ApiResponseTypeProvider.cs,233). That code was originally introduced to resolve https://github.com/dotnet/aspnetcore/issues/56975 so any fixes would have to make sure we don't regress that scenario.

For now, you can workaround this by using the `ProducesAttribute` to inject the actual type. It doesn't implement the interface that the check above is susceptible to:

```csharp
using System.Reflection;
using Microsoft.AspNetCore.Http.Metadata;
using Microsoft.AspNetCore.Mvc;

var bld = WebApplication.CreateBuilder(args);
bld.Services.AddOpenApi();

var app = bld.Build();

app.MapGet("broken", () => new MyResponse { Content = "hello!" })
   .Produces<MyResponse>(200, "application/json"); 

app.MapGet("works", [Produces(typeof(MyResponse))] () => new MyResponse { Content = "hello!" })
   .Produces<MyResponse>(200, "application/json"); 

app.MapGet("also-works", () => new MyResponse2 { Content = "hello!" })
   .Produces<MyResponse2>(200, "application/json"); 



app.MapOpenApi();

app.Run();

public class MyResponse2 : IResult, IEndpointMetadataProvider 
{
    public required string Content { get; set; }

    public static void PopulateMetadata(MethodInfo method, EndpointBuilder builder)
    {
        builder.Metadata.Add(new ProducesAttribute(typeof(MyResponse)));
    }

    public Task ExecuteAsync(HttpContext httpContext)
    {
        return Task.CompletedTask;
    }
}

public class MyResponse : IResult //commenting out IResult or switching to net8.0 sdk fixes it
{
    public required string Content { get; set; }

    public Task ExecuteAsync(HttpContext httpContext)
    {
        return Task.CompletedTask;
    }
}
```

**Comment by dj-nitehawk** [comment]
@captainsafia thanks for the fast response and workaround. however, consider the following scenario:

```cs
app.MapGet("test", () => new MyResponse { Content = "hello!" })
   .Produces<MyResponse>(200, "application/json")
   .WithMetadata( new ProducesAttribute(typeof(ErrorResponse)) { ContentTypes = ["application/problem+json"] });
```

above produces the following output with `AspNetCore.OpenApi`:

```json
{
  "responses": {
    "200": {
      "content": {
        "application/json": {
          "schema": {
            "$ref": "#/components/schemas/MyResponse"
          }
        },
        "application/problem+json": {
          "schema": {
            "$ref": "#/components/schemas/MyResponse"
          }
        }
      }
    }
  }
}
```

first limitation is `ProducesAttribute` can't represent anything other than a 200 status code, so it gets lumped in with the 200 responses, and the entry seems to be pointing to the wrong schema whereas it should $ref to `ErrorResponse`. also this approach has no effect in nswag. `ProducesAttribute` metadata entry is totally ignored in nswag.

**Comment by captainsafia** [comment]
> first limitation is ProducesAttribute can't represent anything other than a 200 status code, so it gets lumped in with the 200 responses, and the entry seems to be pointing to the wrong schema whereas it should $ref to ErrorResponse.

@dj-nitehawk Assuming that `ErrorResponse` is also a type that implements `IResult` you can implement an `IEndpointMetadataProvider` there but use the `ProducesResponseTypeAttribute`. I made [this gist](https://gist.github.com/captainsafia/ef8c957d410b76f4e8a17403964329fd) with the code I used and the output that I got.

> also this approach has no effect in nswag. ProducesAttribute metadata entry is totally ignored in nswag.

I was able to repro this behavior. For the example above, both `Microsoft.AspNetCore.OpenApi` and `Swashbuckle.AspNetCore` produce similar and accurate documents but NSwag doesn't appear to respect the content-type on the `ErrorResponse` at all. 😞 

This one might require further investigation. NSwag's OpenAPI implementation is unique amongst the trio because it uses it's own Open API base types instead of [Microsoft.OpenApi](https://www.nuget.org/packages/Microsoft.OpenApi) and from prior investigations there's a more post-processing that NSwag does on ApiExplorer metadata than what Swashbuckle and Microsoft.AspNetCore.OpenApi do so the issue might be a result of one of those things (or something else entirely because...computers).

**Comment by dj-nitehawk** [comment]
@captainsafia dang 😢
unfortunately for us over at fastendpoints, we're currently tied to nswag. will workaround this for the time being with more postprocessing i suppose. this comes in to play for us because our own [ProblemDetails](https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/DTOs/ProblemDetails.cs) class implements `IResult`. the library automatically adds the 400 produces metadata if there's a fluentvalidator attached to the endpoint.

cheers!

**Comment by captainsafia** [comment]
@dj-nitehawk Took another look at this and it seems like the best option to address both issues here is adding the existing `typeof(IResult)` check to `OpenApiGenerator` and adding additional logic to only skip ASP.NET Core-owned `IResult` implementations.

```csharp
if (typeof(IResult).IsAssignableFrom(metadata.Type) &&
  (typeof(Microsoft.AspNetCore.Http.HttpResults.Accepted).Assembly == metadata.Type.Assembly ||
  typeof(IResult).Assembly == metadata.Type.Assembly))
```

I noticed that your implementation uses `IEndpointMetadataProvider` which is great! A lot of places in our code assume that `Result` implementing types will indicate what metadata they should generate via the `IEndpointMetadataProvider` interface, for example:

https://github.com/dotnet/aspnetcore/blob/d7130a2c5a99147005744969aab7f837c82659d5/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1039-L1040

However, the gotcha here is that the `ProducesResponseTypeMetadata` that you insert is itself an `IResult`-implementing type. The check above would exempt `FastEndpoints`s `ProblemDetails` type from this check (and also everybody else's) from this behavior which should resolve the issue.

I think this bug slipped through because of the assumption the framework makes above that `Result` with no `IEndpointMetadataProvider` means no metadata gets populated and that no implementations of `IEndpointMetadataProvider` in ASP.NET Core themselves insert an IResult-based type into metadata.

The categories impacted here are:

- IResult directly
- Types in ASP.NET Core that implement IResult but not IEndpointMetadataProvider
- Types in ASP.NET Core that implement IResult _and_IEndpointMetadataProvider
- Types in other frameworks that implement IResult but not IEndpointMetadataProvider
- Types in other frameworks that implement IResult _and_ IEndpointMetadataProvider
- Types in other frameworks that implement IResult _and_ IEndpointMetadataProvider _that implements an IResults-implementing type_ (not covered in our tests)

**Comment by sandves** [comment]
Any updates on this? Would really appreciate a fix 🙂

**Comment by captainsafia** [comment]
@sander1095 If you're up for it, here's another bug I'd love your help on.

We made some changes in the ApiResponseTypeProvider in .NET 9 as part of the effort to make the new OpenAPI support native AOT compatible but this regressed behavior in the OpenApiGenerator codepath that the old "WithOpenApi" method used.

If possible, it would be good to explore a tactical change that we can include in a servicing release but if the change is to expansive we can stick with fixing it for .NET 10 and beyond.

**Comment by sander1095** [comment]
Hi @captainsafia ! I wouldn't mind having a look, but the length of this issue and the several topics/projects it mentions makes it difficult to understand the full picture. To avoid creating a PR that makes unnecessary/wrong changes, could you provide some more details:

- About the issue we're fixing here (Adding a code example would be nice)
- A "before/after" code example and OpenAPI document result that shows what scenario needs to be supported after this issue is fixed.

For example, is this issue related specifically to the OpenAPI package? Or also the ApiExplorer? How about Controllers and Minimal APIs? :)

My goal is to understand this issue exactly so I can start by creating some test cases for affected areas.


-----

My guess so far are [your latest comments](https://github.com/dotnet/aspnetcore/issues/57801#issuecomment-2445467057) which mentions changes that need to be made to: `OpenApiGenerator`'s `IResult` logic:

https://github.com/dotnet/aspnetcore/blob/b711334da1b465712860b3017c43fdd27f5a7f2a/src/OpenApi/src/Services/OpenApiGenerator.cs#L109-L112

But when I search for more similar code, I find:

- `ApiResponseTypeProvider.ReadResponseMetadata`
  - https://github.com/dotnet/aspnetcore/blob/b711334da1b465712860b3017c43fdd27f5a7f2a/src/Mvc/Mvc.ApiExplorer/src/ApiResponseTypeProvider.cs#L234-L239
- `RequestDelegateFactory.AddResponseWritingToMethodCall`
  - https://github.com/dotnet/aspnetcore/blob/b711334da1b465712860b3017c43fdd27f5a7f2a/src/Http/Http.Extensions/src/RequestDelegateFactory.cs#L1040-L1044
  - There's more places where `IResult` and `IsAssignableFrom` checks are done in this file
- `EndpointMetadataApiDescriptionProvider.AddSupportedResponseTypes`
  - https://github.com/dotnet/aspnetcore/blob/b711334da1b465712860b3017c43fdd27f5a7f2a/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs#L335-L341

Understanding the issue and knowing _roughly_ what kind of changes need to be made and where (and what changes should NOT be made) should make it easier for me to have a deeper look into this.

**Comment by Junjun-zhao** [comment]
Hi @captainsafia , just bringing this to your attention.  Are there any updates or plans regarding this issue?

**Comment by captainsafia** [comment]
@Junjun-zhao I'm hoping to land this in .NET 10.

**Comment by mitchdenny** [comment]
@captainsafia moved this to RC1 since you mentioned hoping to get to it in 10. Move to backlog if you don't think its going to make it.

**Comment by JobaDiniz** [comment]
I'm using aspnet 8 and I expect to use `Produces<T>` and the openapi is generated based on that - and **not** based on the type returned from the delegate. Why doesn't that work?!?!

**Comment by Junjun-zhao** [comment]
Hi @captainsafia , could you please help check if the fix has been merged into the latest RC1 build #dotnet-sdk-10.0.100-rc.1.25416.107?  We verified https://github.com/dotnet/aspnetcore/issues/57343 that we reported and dup of this one, it still reproduces.

**Comment by captainsafia** [comment]
@Junjun-zhao I couldn't find the commits in that version but they should be in 10.0.100-rc.1.25418.115. The test app should target the net10 framework for this fix.

**Comment by Junjun-zhao** [comment]
Thanks @captainsafia. We tried with build 10.0.100-rc.1.25418.116 and retargeted the app to .NET 10, https://github.com/dotnet/aspnetcore/issues/57343 still reproduces. Could you please take a look at it again? Thanks.

**Comment by captainsafia** [comment]
@Junjun-zhao Can you upload a repro application to GitHub? I made my own repro based on the details of the original issue but there might be something I missed.

**Comment by Junjun-zhao** [comment]
Hi @captainsafia, please try with this demo [ApiDemo_Net10.zip](https://github.com/user-attachments/files/21932518/ApiDemo_Net10.zip)

We checked the latest [CleanArchitecture](https://github.com/ardalis/CleanArchitecture) version, the issue does not repro as it has upgrade to use FastEndpoints.Swagge 6.2.0.  This issue reproduces with FastEndpoints.Swagger 5.28.0.5-beta.

**Comment by Junjun-zhao** [comment]
Hi @captainsafia Gentle ping.  Any updates on this?

**Comment by captainsafia** [comment]
@Junjun-zhao If the issue only repros on a beta version that is over a year old, I'm less inclined to view it as an issue on our end. Does it repro on the latest 7.0.1 versions of the package?

**Comment by Junjun-zhao** [comment]
@captainsafia No. It does not repro on the latest 7.0.1 version.
